// rclnodejs/generated/motion_msgs/motion_msgs__msg__MotionCtrl.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const MotionCtrlRefStruct = StructType({
  cmd_id: primitiveTypes.int32,
  value: primitiveTypes.float64,
});
const MotionCtrlRefArray = ArrayType(MotionCtrlRefStruct);
const MotionCtrlRefStructArray = StructType({
  data: MotionCtrlRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class MotionCtrlWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new MotionCtrlRefStruct(other._refObject.toObject());
      this._cmd_idIntialized = true;
      this._valueIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new MotionCtrlRefStruct();
    this._cmd_idIntialized = false;
    this._valueIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new MotionCtrlWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new MotionCtrlArrayWrapper;
  }
  static get ArrayType() {
    return MotionCtrlArrayWrapper;
  }
  static get refObjectArrayType() {
    return MotionCtrlRefStructArray
  }
  static get refObjectType() {
    return MotionCtrlRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._cmd_idIntialized) {
        throw new TypeError('Invalid argument: cmd_id in MotionCtrl');
      }
      if (!this._valueIntialized) {
        throw new TypeError('Invalid argument: value in MotionCtrl');
      }
    }
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._cmd_idIntialized = true;
    this._refObject.cmd_id = refObject.cmd_id;
    this._valueIntialized = true;
    this._refObject.value = refObject.value;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
  }
  static destoryRawROS(msg) {
    MotionCtrlWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'motion_msgs', subFolder: 'msg', interfaceName: 'MotionCtrl'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get cmd_id() {
    if (!this._cmd_idIntialized) {
      return undefined;
    }
    return this._refObject.cmd_id;
  }
  set cmd_id(value) {
    this._cmd_idIntialized = true;
    this._refObject.cmd_id = value;
  }
  get value() {
    if (!this._valueIntialized) {
      return undefined;
    }
    return this._refObject.value;
  }
  set value(value) {
    this._valueIntialized = true;
    this._refObject.value = value;
  }
  copyRefObject(refObject) {
    this._refObject = new MotionCtrlRefStruct(refObject.toObject());
    this._cmd_idIntialized = true;
    this._valueIntialized = true;
  }
  copy(other) {
    this._refObject = new MotionCtrlRefStruct(other._refObject.toObject());
    this._cmd_idIntialized = true;
    this._valueIntialized = true;
  }
  static get classType() {
    return MotionCtrlWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"cmd_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"value","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"motion_msgs","type":"MotionCtrl","stringUpperBound":null,"isPrimitiveType":false},"msgName":"MotionCtrl"};
  }
  hasMember(name) {
    let memberNames = ["cmd_id","value"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class MotionCtrlArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof MotionCtrlWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new MotionCtrlWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionCtrlArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to MotionCtrlArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new MotionCtrlRefArray(size);
    this._refObject = new MotionCtrlRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new MotionCtrlWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof MotionCtrlArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "MotionCtrlArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      MotionCtrlWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return MotionCtrlWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return MotionCtrlArrayWrapper;
  }
}
module.exports = MotionCtrlWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "cmd_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "value",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "motion_msgs",
    "type": "MotionCtrl",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "MotionCtrl"
}
*/
