// rclnodejs/generated/motion_msgs/motion_msgs__msg__RobotStatus.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const RobotStatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  ctrl_mode_msg: primitiveTypes.uint8,
  robot_mode_msg: primitiveTypes.uint8,
  error_msg: primitiveTypes.uint32,
  warning_msg: primitiveTypes.uint32,
});
const RobotStatusRefArray = ArrayType(RobotStatusRefStruct);
const RobotStatusRefStructArray = StructType({
  data: RobotStatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class RobotStatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new RobotStatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._ctrl_mode_msgIntialized = true;
      this._robot_mode_msgIntialized = true;
      this._error_msgIntialized = true;
      this._warning_msgIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new RobotStatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._ctrl_mode_msgIntialized = false;
    this._robot_mode_msgIntialized = false;
    this._error_msgIntialized = false;
    this._warning_msgIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new RobotStatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new RobotStatusArrayWrapper;
  }
  static get ArrayType() {
    return RobotStatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return RobotStatusRefStructArray
  }
  static get refObjectType() {
    return RobotStatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._ctrl_mode_msgIntialized) {
        throw new TypeError('Invalid argument: ctrl_mode_msg in RobotStatus');
      }
      if (!this._robot_mode_msgIntialized) {
        throw new TypeError('Invalid argument: robot_mode_msg in RobotStatus');
      }
      if (!this._error_msgIntialized) {
        throw new TypeError('Invalid argument: error_msg in RobotStatus');
      }
      if (!this._warning_msgIntialized) {
        throw new TypeError('Invalid argument: warning_msg in RobotStatus');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._ctrl_mode_msgIntialized = true;
    this._refObject.ctrl_mode_msg = refObject.ctrl_mode_msg;
    this._robot_mode_msgIntialized = true;
    this._refObject.robot_mode_msg = refObject.robot_mode_msg;
    this._error_msgIntialized = true;
    this._refObject.error_msg = refObject.error_msg;
    this._warning_msgIntialized = true;
    this._refObject.warning_msg = refObject.warning_msg;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    RobotStatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'motion_msgs', subFolder: 'msg', interfaceName: 'RobotStatus'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get ctrl_mode_msg() {
    if (!this._ctrl_mode_msgIntialized) {
      return undefined;
    }
    return this._refObject.ctrl_mode_msg;
  }
  set ctrl_mode_msg(value) {
    this._ctrl_mode_msgIntialized = true;
    this._refObject.ctrl_mode_msg = value;
  }
  get robot_mode_msg() {
    if (!this._robot_mode_msgIntialized) {
      return undefined;
    }
    return this._refObject.robot_mode_msg;
  }
  set robot_mode_msg(value) {
    this._robot_mode_msgIntialized = true;
    this._refObject.robot_mode_msg = value;
  }
  get error_msg() {
    if (!this._error_msgIntialized) {
      return undefined;
    }
    return this._refObject.error_msg;
  }
  set error_msg(value) {
    this._error_msgIntialized = true;
    this._refObject.error_msg = value;
  }
  get warning_msg() {
    if (!this._warning_msgIntialized) {
      return undefined;
    }
    return this._refObject.warning_msg;
  }
  set warning_msg(value) {
    this._warning_msgIntialized = true;
    this._refObject.warning_msg = value;
  }
  copyRefObject(refObject) {
    this._refObject = new RobotStatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._ctrl_mode_msgIntialized = true;
    this._robot_mode_msgIntialized = true;
    this._error_msgIntialized = true;
    this._warning_msgIntialized = true;
  }
  copy(other) {
    this._refObject = new RobotStatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._ctrl_mode_msgIntialized = true;
    this._robot_mode_msgIntialized = true;
    this._error_msgIntialized = true;
    this._warning_msgIntialized = true;
  }
  static get classType() {
    return RobotStatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"ctrl_mode_msg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"robot_mode_msg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"error_msg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"warning_msg","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"motion_msgs","type":"RobotStatus","stringUpperBound":null,"isPrimitiveType":false},"msgName":"RobotStatus"};
  }
  hasMember(name) {
    let memberNames = ["header","ctrl_mode_msg","robot_mode_msg","error_msg","warning_msg"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class RobotStatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof RobotStatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new RobotStatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotStatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to RobotStatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new RobotStatusRefArray(size);
    this._refObject = new RobotStatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new RobotStatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof RobotStatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "RobotStatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      RobotStatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return RobotStatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return RobotStatusArrayWrapper;
  }
}
module.exports = RobotStatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "ctrl_mode_msg",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "robot_mode_msg",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "error_msg",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "warning_msg",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "motion_msgs",
    "type": "RobotStatus",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "RobotStatus"
}
*/
