// rclnodejs/generated/motion_msgs/motion_msgs__msg__LegMotors.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const LegMotorsRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  left_hip_enc_rev: primitiveTypes.int32,
  left_hip_pos: primitiveTypes.float64,
  left_hip_vel: primitiveTypes.float64,
  left_hip_iq: primitiveTypes.float64,
  left_knee_enc_rev: primitiveTypes.int32,
  left_knee_pos: primitiveTypes.float64,
  left_knee_vel: primitiveTypes.float64,
  left_knee_iq: primitiveTypes.float64,
  left_wheel_enc_rev: primitiveTypes.int32,
  left_wheel_pos: primitiveTypes.float64,
  left_wheel_vel: primitiveTypes.float64,
  left_wheel_iq: primitiveTypes.float64,
  right_hip_enc_rev: primitiveTypes.int32,
  right_hip_pos: primitiveTypes.float64,
  right_hip_vel: primitiveTypes.float64,
  right_hip_iq: primitiveTypes.float64,
  right_knee_enc_rev: primitiveTypes.int32,
  right_knee_pos: primitiveTypes.float64,
  right_knee_vel: primitiveTypes.float64,
  right_knee_iq: primitiveTypes.float64,
  right_wheel_enc_rev: primitiveTypes.int32,
  right_wheel_pos: primitiveTypes.float64,
  right_wheel_vel: primitiveTypes.float64,
  right_wheel_iq: primitiveTypes.float64,
  left_leg_length: primitiveTypes.float64,
  right_leg_length: primitiveTypes.float64,
});
const LegMotorsRefArray = ArrayType(LegMotorsRefStruct);
const LegMotorsRefStructArray = StructType({
  data: LegMotorsRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class LegMotorsWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new LegMotorsRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._left_hip_enc_revIntialized = true;
      this._left_hip_posIntialized = true;
      this._left_hip_velIntialized = true;
      this._left_hip_iqIntialized = true;
      this._left_knee_enc_revIntialized = true;
      this._left_knee_posIntialized = true;
      this._left_knee_velIntialized = true;
      this._left_knee_iqIntialized = true;
      this._left_wheel_enc_revIntialized = true;
      this._left_wheel_posIntialized = true;
      this._left_wheel_velIntialized = true;
      this._left_wheel_iqIntialized = true;
      this._right_hip_enc_revIntialized = true;
      this._right_hip_posIntialized = true;
      this._right_hip_velIntialized = true;
      this._right_hip_iqIntialized = true;
      this._right_knee_enc_revIntialized = true;
      this._right_knee_posIntialized = true;
      this._right_knee_velIntialized = true;
      this._right_knee_iqIntialized = true;
      this._right_wheel_enc_revIntialized = true;
      this._right_wheel_posIntialized = true;
      this._right_wheel_velIntialized = true;
      this._right_wheel_iqIntialized = true;
      this._left_leg_lengthIntialized = true;
      this._right_leg_lengthIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new LegMotorsRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._left_hip_enc_revIntialized = false;
    this._left_hip_posIntialized = false;
    this._left_hip_velIntialized = false;
    this._left_hip_iqIntialized = false;
    this._left_knee_enc_revIntialized = false;
    this._left_knee_posIntialized = false;
    this._left_knee_velIntialized = false;
    this._left_knee_iqIntialized = false;
    this._left_wheel_enc_revIntialized = false;
    this._left_wheel_posIntialized = false;
    this._left_wheel_velIntialized = false;
    this._left_wheel_iqIntialized = false;
    this._right_hip_enc_revIntialized = false;
    this._right_hip_posIntialized = false;
    this._right_hip_velIntialized = false;
    this._right_hip_iqIntialized = false;
    this._right_knee_enc_revIntialized = false;
    this._right_knee_posIntialized = false;
    this._right_knee_velIntialized = false;
    this._right_knee_iqIntialized = false;
    this._right_wheel_enc_revIntialized = false;
    this._right_wheel_posIntialized = false;
    this._right_wheel_velIntialized = false;
    this._right_wheel_iqIntialized = false;
    this._left_leg_lengthIntialized = false;
    this._right_leg_lengthIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new LegMotorsWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new LegMotorsArrayWrapper;
  }
  static get ArrayType() {
    return LegMotorsArrayWrapper;
  }
  static get refObjectArrayType() {
    return LegMotorsRefStructArray
  }
  static get refObjectType() {
    return LegMotorsRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._left_hip_enc_revIntialized) {
        throw new TypeError('Invalid argument: left_hip_enc_rev in LegMotors');
      }
      if (!this._left_hip_posIntialized) {
        throw new TypeError('Invalid argument: left_hip_pos in LegMotors');
      }
      if (!this._left_hip_velIntialized) {
        throw new TypeError('Invalid argument: left_hip_vel in LegMotors');
      }
      if (!this._left_hip_iqIntialized) {
        throw new TypeError('Invalid argument: left_hip_iq in LegMotors');
      }
      if (!this._left_knee_enc_revIntialized) {
        throw new TypeError('Invalid argument: left_knee_enc_rev in LegMotors');
      }
      if (!this._left_knee_posIntialized) {
        throw new TypeError('Invalid argument: left_knee_pos in LegMotors');
      }
      if (!this._left_knee_velIntialized) {
        throw new TypeError('Invalid argument: left_knee_vel in LegMotors');
      }
      if (!this._left_knee_iqIntialized) {
        throw new TypeError('Invalid argument: left_knee_iq in LegMotors');
      }
      if (!this._left_wheel_enc_revIntialized) {
        throw new TypeError('Invalid argument: left_wheel_enc_rev in LegMotors');
      }
      if (!this._left_wheel_posIntialized) {
        throw new TypeError('Invalid argument: left_wheel_pos in LegMotors');
      }
      if (!this._left_wheel_velIntialized) {
        throw new TypeError('Invalid argument: left_wheel_vel in LegMotors');
      }
      if (!this._left_wheel_iqIntialized) {
        throw new TypeError('Invalid argument: left_wheel_iq in LegMotors');
      }
      if (!this._right_hip_enc_revIntialized) {
        throw new TypeError('Invalid argument: right_hip_enc_rev in LegMotors');
      }
      if (!this._right_hip_posIntialized) {
        throw new TypeError('Invalid argument: right_hip_pos in LegMotors');
      }
      if (!this._right_hip_velIntialized) {
        throw new TypeError('Invalid argument: right_hip_vel in LegMotors');
      }
      if (!this._right_hip_iqIntialized) {
        throw new TypeError('Invalid argument: right_hip_iq in LegMotors');
      }
      if (!this._right_knee_enc_revIntialized) {
        throw new TypeError('Invalid argument: right_knee_enc_rev in LegMotors');
      }
      if (!this._right_knee_posIntialized) {
        throw new TypeError('Invalid argument: right_knee_pos in LegMotors');
      }
      if (!this._right_knee_velIntialized) {
        throw new TypeError('Invalid argument: right_knee_vel in LegMotors');
      }
      if (!this._right_knee_iqIntialized) {
        throw new TypeError('Invalid argument: right_knee_iq in LegMotors');
      }
      if (!this._right_wheel_enc_revIntialized) {
        throw new TypeError('Invalid argument: right_wheel_enc_rev in LegMotors');
      }
      if (!this._right_wheel_posIntialized) {
        throw new TypeError('Invalid argument: right_wheel_pos in LegMotors');
      }
      if (!this._right_wheel_velIntialized) {
        throw new TypeError('Invalid argument: right_wheel_vel in LegMotors');
      }
      if (!this._right_wheel_iqIntialized) {
        throw new TypeError('Invalid argument: right_wheel_iq in LegMotors');
      }
      if (!this._left_leg_lengthIntialized) {
        throw new TypeError('Invalid argument: left_leg_length in LegMotors');
      }
      if (!this._right_leg_lengthIntialized) {
        throw new TypeError('Invalid argument: right_leg_length in LegMotors');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._left_hip_enc_revIntialized = true;
    this._refObject.left_hip_enc_rev = refObject.left_hip_enc_rev;
    this._left_hip_posIntialized = true;
    this._refObject.left_hip_pos = refObject.left_hip_pos;
    this._left_hip_velIntialized = true;
    this._refObject.left_hip_vel = refObject.left_hip_vel;
    this._left_hip_iqIntialized = true;
    this._refObject.left_hip_iq = refObject.left_hip_iq;
    this._left_knee_enc_revIntialized = true;
    this._refObject.left_knee_enc_rev = refObject.left_knee_enc_rev;
    this._left_knee_posIntialized = true;
    this._refObject.left_knee_pos = refObject.left_knee_pos;
    this._left_knee_velIntialized = true;
    this._refObject.left_knee_vel = refObject.left_knee_vel;
    this._left_knee_iqIntialized = true;
    this._refObject.left_knee_iq = refObject.left_knee_iq;
    this._left_wheel_enc_revIntialized = true;
    this._refObject.left_wheel_enc_rev = refObject.left_wheel_enc_rev;
    this._left_wheel_posIntialized = true;
    this._refObject.left_wheel_pos = refObject.left_wheel_pos;
    this._left_wheel_velIntialized = true;
    this._refObject.left_wheel_vel = refObject.left_wheel_vel;
    this._left_wheel_iqIntialized = true;
    this._refObject.left_wheel_iq = refObject.left_wheel_iq;
    this._right_hip_enc_revIntialized = true;
    this._refObject.right_hip_enc_rev = refObject.right_hip_enc_rev;
    this._right_hip_posIntialized = true;
    this._refObject.right_hip_pos = refObject.right_hip_pos;
    this._right_hip_velIntialized = true;
    this._refObject.right_hip_vel = refObject.right_hip_vel;
    this._right_hip_iqIntialized = true;
    this._refObject.right_hip_iq = refObject.right_hip_iq;
    this._right_knee_enc_revIntialized = true;
    this._refObject.right_knee_enc_rev = refObject.right_knee_enc_rev;
    this._right_knee_posIntialized = true;
    this._refObject.right_knee_pos = refObject.right_knee_pos;
    this._right_knee_velIntialized = true;
    this._refObject.right_knee_vel = refObject.right_knee_vel;
    this._right_knee_iqIntialized = true;
    this._refObject.right_knee_iq = refObject.right_knee_iq;
    this._right_wheel_enc_revIntialized = true;
    this._refObject.right_wheel_enc_rev = refObject.right_wheel_enc_rev;
    this._right_wheel_posIntialized = true;
    this._refObject.right_wheel_pos = refObject.right_wheel_pos;
    this._right_wheel_velIntialized = true;
    this._refObject.right_wheel_vel = refObject.right_wheel_vel;
    this._right_wheel_iqIntialized = true;
    this._refObject.right_wheel_iq = refObject.right_wheel_iq;
    this._left_leg_lengthIntialized = true;
    this._refObject.left_leg_length = refObject.left_leg_length;
    this._right_leg_lengthIntialized = true;
    this._refObject.right_leg_length = refObject.right_leg_length;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    LegMotorsWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'motion_msgs', subFolder: 'msg', interfaceName: 'LegMotors'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get left_hip_enc_rev() {
    if (!this._left_hip_enc_revIntialized) {
      return undefined;
    }
    return this._refObject.left_hip_enc_rev;
  }
  set left_hip_enc_rev(value) {
    this._left_hip_enc_revIntialized = true;
    this._refObject.left_hip_enc_rev = value;
  }
  get left_hip_pos() {
    if (!this._left_hip_posIntialized) {
      return undefined;
    }
    return this._refObject.left_hip_pos;
  }
  set left_hip_pos(value) {
    this._left_hip_posIntialized = true;
    this._refObject.left_hip_pos = value;
  }
  get left_hip_vel() {
    if (!this._left_hip_velIntialized) {
      return undefined;
    }
    return this._refObject.left_hip_vel;
  }
  set left_hip_vel(value) {
    this._left_hip_velIntialized = true;
    this._refObject.left_hip_vel = value;
  }
  get left_hip_iq() {
    if (!this._left_hip_iqIntialized) {
      return undefined;
    }
    return this._refObject.left_hip_iq;
  }
  set left_hip_iq(value) {
    this._left_hip_iqIntialized = true;
    this._refObject.left_hip_iq = value;
  }
  get left_knee_enc_rev() {
    if (!this._left_knee_enc_revIntialized) {
      return undefined;
    }
    return this._refObject.left_knee_enc_rev;
  }
  set left_knee_enc_rev(value) {
    this._left_knee_enc_revIntialized = true;
    this._refObject.left_knee_enc_rev = value;
  }
  get left_knee_pos() {
    if (!this._left_knee_posIntialized) {
      return undefined;
    }
    return this._refObject.left_knee_pos;
  }
  set left_knee_pos(value) {
    this._left_knee_posIntialized = true;
    this._refObject.left_knee_pos = value;
  }
  get left_knee_vel() {
    if (!this._left_knee_velIntialized) {
      return undefined;
    }
    return this._refObject.left_knee_vel;
  }
  set left_knee_vel(value) {
    this._left_knee_velIntialized = true;
    this._refObject.left_knee_vel = value;
  }
  get left_knee_iq() {
    if (!this._left_knee_iqIntialized) {
      return undefined;
    }
    return this._refObject.left_knee_iq;
  }
  set left_knee_iq(value) {
    this._left_knee_iqIntialized = true;
    this._refObject.left_knee_iq = value;
  }
  get left_wheel_enc_rev() {
    if (!this._left_wheel_enc_revIntialized) {
      return undefined;
    }
    return this._refObject.left_wheel_enc_rev;
  }
  set left_wheel_enc_rev(value) {
    this._left_wheel_enc_revIntialized = true;
    this._refObject.left_wheel_enc_rev = value;
  }
  get left_wheel_pos() {
    if (!this._left_wheel_posIntialized) {
      return undefined;
    }
    return this._refObject.left_wheel_pos;
  }
  set left_wheel_pos(value) {
    this._left_wheel_posIntialized = true;
    this._refObject.left_wheel_pos = value;
  }
  get left_wheel_vel() {
    if (!this._left_wheel_velIntialized) {
      return undefined;
    }
    return this._refObject.left_wheel_vel;
  }
  set left_wheel_vel(value) {
    this._left_wheel_velIntialized = true;
    this._refObject.left_wheel_vel = value;
  }
  get left_wheel_iq() {
    if (!this._left_wheel_iqIntialized) {
      return undefined;
    }
    return this._refObject.left_wheel_iq;
  }
  set left_wheel_iq(value) {
    this._left_wheel_iqIntialized = true;
    this._refObject.left_wheel_iq = value;
  }
  get right_hip_enc_rev() {
    if (!this._right_hip_enc_revIntialized) {
      return undefined;
    }
    return this._refObject.right_hip_enc_rev;
  }
  set right_hip_enc_rev(value) {
    this._right_hip_enc_revIntialized = true;
    this._refObject.right_hip_enc_rev = value;
  }
  get right_hip_pos() {
    if (!this._right_hip_posIntialized) {
      return undefined;
    }
    return this._refObject.right_hip_pos;
  }
  set right_hip_pos(value) {
    this._right_hip_posIntialized = true;
    this._refObject.right_hip_pos = value;
  }
  get right_hip_vel() {
    if (!this._right_hip_velIntialized) {
      return undefined;
    }
    return this._refObject.right_hip_vel;
  }
  set right_hip_vel(value) {
    this._right_hip_velIntialized = true;
    this._refObject.right_hip_vel = value;
  }
  get right_hip_iq() {
    if (!this._right_hip_iqIntialized) {
      return undefined;
    }
    return this._refObject.right_hip_iq;
  }
  set right_hip_iq(value) {
    this._right_hip_iqIntialized = true;
    this._refObject.right_hip_iq = value;
  }
  get right_knee_enc_rev() {
    if (!this._right_knee_enc_revIntialized) {
      return undefined;
    }
    return this._refObject.right_knee_enc_rev;
  }
  set right_knee_enc_rev(value) {
    this._right_knee_enc_revIntialized = true;
    this._refObject.right_knee_enc_rev = value;
  }
  get right_knee_pos() {
    if (!this._right_knee_posIntialized) {
      return undefined;
    }
    return this._refObject.right_knee_pos;
  }
  set right_knee_pos(value) {
    this._right_knee_posIntialized = true;
    this._refObject.right_knee_pos = value;
  }
  get right_knee_vel() {
    if (!this._right_knee_velIntialized) {
      return undefined;
    }
    return this._refObject.right_knee_vel;
  }
  set right_knee_vel(value) {
    this._right_knee_velIntialized = true;
    this._refObject.right_knee_vel = value;
  }
  get right_knee_iq() {
    if (!this._right_knee_iqIntialized) {
      return undefined;
    }
    return this._refObject.right_knee_iq;
  }
  set right_knee_iq(value) {
    this._right_knee_iqIntialized = true;
    this._refObject.right_knee_iq = value;
  }
  get right_wheel_enc_rev() {
    if (!this._right_wheel_enc_revIntialized) {
      return undefined;
    }
    return this._refObject.right_wheel_enc_rev;
  }
  set right_wheel_enc_rev(value) {
    this._right_wheel_enc_revIntialized = true;
    this._refObject.right_wheel_enc_rev = value;
  }
  get right_wheel_pos() {
    if (!this._right_wheel_posIntialized) {
      return undefined;
    }
    return this._refObject.right_wheel_pos;
  }
  set right_wheel_pos(value) {
    this._right_wheel_posIntialized = true;
    this._refObject.right_wheel_pos = value;
  }
  get right_wheel_vel() {
    if (!this._right_wheel_velIntialized) {
      return undefined;
    }
    return this._refObject.right_wheel_vel;
  }
  set right_wheel_vel(value) {
    this._right_wheel_velIntialized = true;
    this._refObject.right_wheel_vel = value;
  }
  get right_wheel_iq() {
    if (!this._right_wheel_iqIntialized) {
      return undefined;
    }
    return this._refObject.right_wheel_iq;
  }
  set right_wheel_iq(value) {
    this._right_wheel_iqIntialized = true;
    this._refObject.right_wheel_iq = value;
  }
  get left_leg_length() {
    if (!this._left_leg_lengthIntialized) {
      return undefined;
    }
    return this._refObject.left_leg_length;
  }
  set left_leg_length(value) {
    this._left_leg_lengthIntialized = true;
    this._refObject.left_leg_length = value;
  }
  get right_leg_length() {
    if (!this._right_leg_lengthIntialized) {
      return undefined;
    }
    return this._refObject.right_leg_length;
  }
  set right_leg_length(value) {
    this._right_leg_lengthIntialized = true;
    this._refObject.right_leg_length = value;
  }
  copyRefObject(refObject) {
    this._refObject = new LegMotorsRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._left_hip_enc_revIntialized = true;
    this._left_hip_posIntialized = true;
    this._left_hip_velIntialized = true;
    this._left_hip_iqIntialized = true;
    this._left_knee_enc_revIntialized = true;
    this._left_knee_posIntialized = true;
    this._left_knee_velIntialized = true;
    this._left_knee_iqIntialized = true;
    this._left_wheel_enc_revIntialized = true;
    this._left_wheel_posIntialized = true;
    this._left_wheel_velIntialized = true;
    this._left_wheel_iqIntialized = true;
    this._right_hip_enc_revIntialized = true;
    this._right_hip_posIntialized = true;
    this._right_hip_velIntialized = true;
    this._right_hip_iqIntialized = true;
    this._right_knee_enc_revIntialized = true;
    this._right_knee_posIntialized = true;
    this._right_knee_velIntialized = true;
    this._right_knee_iqIntialized = true;
    this._right_wheel_enc_revIntialized = true;
    this._right_wheel_posIntialized = true;
    this._right_wheel_velIntialized = true;
    this._right_wheel_iqIntialized = true;
    this._left_leg_lengthIntialized = true;
    this._right_leg_lengthIntialized = true;
  }
  copy(other) {
    this._refObject = new LegMotorsRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._left_hip_enc_revIntialized = true;
    this._left_hip_posIntialized = true;
    this._left_hip_velIntialized = true;
    this._left_hip_iqIntialized = true;
    this._left_knee_enc_revIntialized = true;
    this._left_knee_posIntialized = true;
    this._left_knee_velIntialized = true;
    this._left_knee_iqIntialized = true;
    this._left_wheel_enc_revIntialized = true;
    this._left_wheel_posIntialized = true;
    this._left_wheel_velIntialized = true;
    this._left_wheel_iqIntialized = true;
    this._right_hip_enc_revIntialized = true;
    this._right_hip_posIntialized = true;
    this._right_hip_velIntialized = true;
    this._right_hip_iqIntialized = true;
    this._right_knee_enc_revIntialized = true;
    this._right_knee_posIntialized = true;
    this._right_knee_velIntialized = true;
    this._right_knee_iqIntialized = true;
    this._right_wheel_enc_revIntialized = true;
    this._right_wheel_posIntialized = true;
    this._right_wheel_velIntialized = true;
    this._right_wheel_iqIntialized = true;
    this._left_leg_lengthIntialized = true;
    this._right_leg_lengthIntialized = true;
  }
  static get classType() {
    return LegMotorsWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"left_hip_enc_rev","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_hip_pos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_hip_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_hip_iq","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_knee_enc_rev","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_knee_pos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_knee_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_knee_iq","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_wheel_enc_rev","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_wheel_pos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_wheel_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_wheel_iq","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_hip_enc_rev","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_hip_pos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_hip_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_hip_iq","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_knee_enc_rev","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_knee_pos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_knee_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_knee_iq","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_wheel_enc_rev","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_wheel_pos","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_wheel_vel","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_wheel_iq","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"left_leg_length","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"right_leg_length","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"motion_msgs","type":"LegMotors","stringUpperBound":null,"isPrimitiveType":false},"msgName":"LegMotors"};
  }
  hasMember(name) {
    let memberNames = ["header","left_hip_enc_rev","left_hip_pos","left_hip_vel","left_hip_iq","left_knee_enc_rev","left_knee_pos","left_knee_vel","left_knee_iq","left_wheel_enc_rev","left_wheel_pos","left_wheel_vel","left_wheel_iq","right_hip_enc_rev","right_hip_pos","right_hip_vel","right_hip_iq","right_knee_enc_rev","right_knee_pos","right_knee_vel","right_knee_iq","right_wheel_enc_rev","right_wheel_pos","right_wheel_vel","right_wheel_iq","left_leg_length","right_leg_length"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class LegMotorsArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof LegMotorsWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new LegMotorsWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LegMotorsArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to LegMotorsArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new LegMotorsRefArray(size);
    this._refObject = new LegMotorsRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new LegMotorsWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof LegMotorsArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "LegMotorsArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      LegMotorsWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return LegMotorsWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return LegMotorsArrayWrapper;
  }
}
module.exports = LegMotorsWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "left_hip_enc_rev",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_hip_pos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_hip_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_hip_iq",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_knee_enc_rev",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_knee_pos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_knee_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_knee_iq",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_wheel_enc_rev",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_wheel_pos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_wheel_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_wheel_iq",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_hip_enc_rev",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_hip_pos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_hip_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_hip_iq",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_knee_enc_rev",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_knee_pos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_knee_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_knee_iq",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_wheel_enc_rev",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_wheel_pos",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_wheel_vel",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_wheel_iq",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "left_leg_length",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "right_leg_length",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "motion_msgs",
    "type": "LegMotors",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "LegMotors"
}
*/
